packages/flutter_tools/lib/src/emulator.dart

```diff
@@ -0,0 +1,176 @@
+        exactlyMatchesEmulatorId, orElse: () => null);
+    if (exactMatch != null) {
+      yield exactMatch;
+      return;
+    }
+
+    // Match on a id or name starting with [emulatorId].
+    for (Emulator emulator in emulators.where(startsWithEmulatorId))
+      yield emulator;
+  }
+
+  /// Return the list of available emulators, filtered by any user-specified emulator id.
+  Stream<Emulator> getEmulators() {
+    return hasSpecifiedEmulatorId
+        ? getEmulatorsById(specifiedEmulatorId)
+        : getAllAvailableEmulators();
+  }
+
+  Iterable<EmulatorDiscovery> get _platformDiscoverers {
+    return _emulatorDiscoverers.where((EmulatorDiscovery discoverer) => discoverer.supportsPlatform);
+  }
+
+  /// Return the list of all connected emulators.
+  Stream<Emulator> getAllAvailableEmulators() async* {
+    for (EmulatorDiscovery discoverer in _platformDiscoverers) {
+      for (Emulator emulator in await discoverer.emulators) {
+        yield emulator;
+      }
+    }
+  }
+
+  /// Whether we're capable of listing any emulators given the current environment configuration.
+  bool get canListAnything {
+    return _platformDiscoverers.any((EmulatorDiscovery discoverer) => discoverer.canListAnything);
+  }
+
+  /// Get diagnostics about issues with any emulators.
+  Future<List<String>> getEmulatorDiagnostics() async {
+    final List<String> diagnostics = <String>[];
+    for (EmulatorDiscovery discoverer in _platformDiscoverers) {
+      diagnostics.addAll(await discoverer.getDiagnostics());
+    }
+    return diagnostics;
+  }
+}
+
+/// An abstract class to discover and enumerate a specific type of emulators.
+abstract class EmulatorDiscovery {
+  bool get supportsPlatform;
+
+  /// Whether this emulator discovery is capable of listing any emulators given the
+  /// current environment configuration.
+  bool get canListAnything;
+
+  Future<List<Emulator>> get emulators;
+
+  /// Gets a list of diagnostic messages pertaining to issues with any available
+  /// emulators (will be an empty list if there are no issues).
+  Future<List<String>> getDiagnostics() => new Future<List<String>>.value(<String>[]);
+}
+
+abstract class Emulator {
+  Emulator(this.id, this.hasConfig);
+
+  final String id;
+  final bool hasConfig;
+  String get name;
+  String get manufacturer;
+  String get label;
+
+  @override
+  int get hashCode => id.hashCode;
+
+  @override
+  bool operator ==(dynamic other) {
+    if (identical(this, other))
+      return true;
+    if (other is! Emulator)
+      return false;
+    return id == other.id;
+  }
+
+  void launch();
+
+  @override
+  String toString() => name;
+
+  static Stream<String> descriptions(List<Emulator> emulators) async* {
+    if (emulators.isEmpty)
+      return;
+
+    // Extract emulators information
+    final List<List<String>> table = <List<String>>[];
+    for (Emulator emulator in emulators) {
+      table.add(<String>[
+        emulator.name ?? emulator.id ?? '',
```

Comment:
It looks like we always have IDs, but not always names. Perhaps:

id (name) • manufacturer • label

What is label (vs name)?

Reply:
I thought label could be used for additional info that might vary between platforms (eg. architecture or other stuff we might want to build into a string but not give its own column as it's not consistent between Android/iOS).

The ID is the unique thing a user can use to launch an emulator regardless of how unique the name is, so I think it should be in its own column (however it does look a bit naff).

If we put `id (name)` in the first column, it'd could look like `Nexus_5X_API_27_X86 (Nexus 5X)` which seems a bit of a moutful. WDYT?

Reply:
Ugh, apparently Cmd+Enter starts a review! Why would I want to review my own changes? 😆

Reply:
I've gone with the first for now, so ID is the first column. We can always tweak it further if we think it looks weird with similar name/IDs.

Reply:
👍